istioctl/pkg/validate/validate.go

```diff
@@ -186,7 +186,7 @@ func (v *validator) validateDeploymentLabel(istioNamespace string, un *unstructu
 	if un.GetNamespace() == handleNamespace(istioNamespace) {
 		return
 	}
-	labels := un.GetLabels()
+	labels := GetTemplateLabels(un)
 	for _, l := range istioDeploymentLabel {
 		if _, ok := labels[l]; !ok {
 			log.Warnf("deployment %q may not provide Istio metrics and telemetry without label %q."+
```

Comment:
Because we are logging the output it looks silly:

```
2020-09-15T13:42:50.407866Z	warn	deployment "httpbin/:" may not provide Istio metrics and telemetry without label "version". See https://istio.io/v1.8/docs/ops/deployment/requirements/
"samples/httpbin/httpbin.yaml" is valid
```

Either pass the `writer` that `validateFiles()` uses deeper into the methods; or modify `validateFile()` to return two values, errors and warnings, and have validateFiles() send the new warning return to its `writer`.  The second option seems more re-usable.

Reply:
Let me try.

Reply:
I'm not sure if I understand it correctly, but I tried passing `writer` to  `validateFile` which internally calls `validateResource` > `validateDeploymentLabel` and print the msg to `writer`.  Which outputs the following.

```
Validating samples/httpbin/sample-client/fortio-deploy.yaml...
deployment "fortio-deploy/:" may not provide Istio metrics and telemetry without label "version". See https://istio.io/v1.8/docs/ops/deployment/requirements/
"samples/httpbin/sample-client/fortio-deploy.yaml" is valid
```

I think the current implementation with `log.Warnf` still looks reasonable and capture user attention to what went wrong.

```
$ istioctl validate -f samples/httpbin/sample-client/fortio-deploy.yaml

2020-09-16T07:18:41.180277Z     warn    deployment "fortio-deploy/:" may not provide Istio metrics and telemetry without label "app". See https://istio.io/latest/docs/ops/deployment/requirements/
2020-09-16T07:18:41.180305Z     warn    deployment "fortio-deploy/:" may not provide Istio metrics and telemetry without label "version". See https://istio.io/latest/docs/ops/deployment/requirements/
"samples/httpbin/sample-client/fortio-deploy.yaml" is valid
```
AFAI understood the purpose of `validateDeploymentLabel` is to just check and warn if `app` and `version` labels are missing and left other validation for `validateResource`?

Reply:
The message is fine.  We don't like the log timestamps.
If you pass in writer, which I think comes from cmd.OutOrStderr() (or perhaps cmd.OutOrStdout()), the unit tests can easily check that the warnings appear.  If you write the warnings with log.Warnf() not only is the timestamp there, but the unit tests don't see the warnings.

Reply:
Make sense. Updated to use writer which comes from `c.OutOrStderr()` from `validateFiles`. will update test accordingly.